Executor Runtimes

Piszmog opened the pull request:
Closes #46175 and #46176.

_Note: Most of the new lines are from adding Unit Tests. Over 120 new unit tests were added._

This PR will help position us into a situation where Executors can understand the environment they are deployed to and handle Jobs with that specific environment.

So for example if docker has been configured, the executor will handle all jobs with docker.

In the future, we will add runtimes for Firecracker, K8s, and more.

## Changes

- Split `command` into two packages
    - `worker/command` and `worker/runner`
    - This is to help with getting test coverage and to make integrating the `worker/runtime` package more easily
- General housekeeping (fixing function names, remove variable shadowing when not necessary)
- Created an interface `Runtime` in the executor app
    - This interface will prepare the workspace, create a runner (what will actually execute the steps), and build the commands for the runner to execute
    - At startup, the runtime is configured based on how the host is configured.
        - At the moment, I commented out this setup call to be backwards compatible 
    - When a `Job` is processed, the configured runtime is used to execute the `Job`
- Created a `dockerRuntime` struct to handle "docker" runtimes
- Added a special case for `src`
    - If `Job.CLISteps` is set, then we fallback to the "legacy" way
- Over 120 new Unit Tests 

## New Issues
Created the following Issue as I was implementing this PR.

- #47819

## Demo

The following is using the native SSBC for testing.

This just shows that,
* A job is successfully processed
* The logs confirm the commands

### "Legacy" Flow

https://user-images.githubusercontent.com/38407415/222802100-0672b46d-890f-4a5e-82c1-7fd9b13ee44d.mov

### Runtime flow (docker)

https://user-images.githubusercontent.com/38407415/222801920-ad3ec496-64ae-4c71-9467-71ff8e4b98ac.mov

## Test plan

Added unit tests and did functional testing.

eseliger commented on enterprise/internal/executor/types/job.go:
```diff
@@ -67,6 +67,8 @@ type Job struct {
 	// variable.
 	DockerAuthConfig DockerAuthConfig `json:"dockerAuthConfig,omitempty"`
+
+	RuntimeMode RuntimeMode `json:"runtimeMode"`
```
Maybe I'm missing something here, but I think the executor should decide which runtime to use and not the Sourcegraph server - after all, the running executor is the only thing that knows what runtimes it supports on the given host. For example, an existing executor of today could not easily run on k8s without additional information about a cluster to connect to.

efritz commented on enterprise/cmd/executor/internal/worker/worker.go:
```diff
@@ -97,18 +91,34 @@ func NewWorker(observationCtx *observation.Context, nameSet *janitor.NameSet, op
+		ExecutorToken:  options.QueueOptions.BaseClientOptions.EndpointOptions.Token,
+	}
+
+	// Configure the supported runtimes
+	// TODO: uncomment when firecracker runtime is complete
+	//jobRuntime, err := runtime.New(observationCtx.Logger, commandOps, filesClient, options.CommandOptions, cloneOptions)
+	//if err != nil {
+	//	return nil, err
+	//}
```
Do we need this pre-merge?

Piszmog replied:
No (it's complicated). So if I uncomment the code above to allow Docker runtimes to be handled in the new way, Firecracker environments wont work*. Since Firecracker environments are also docker runtimes, if I enable the new runtime stuff without a firecracker implementation, then Firecracker will never be used. So the runtime has be commented out until then.

I could add logic to `runtime.New` to check for Firecracker, but then I'm essentially getting 3 birds stoned at once and that is too much.

efritz commented on enterprise/cmd/executor/internal/worker/runtime/runtime.go:
```diff
@@ -0,0 +1,57 @@
+	cloneOpts workspace.CloneOptions,
+	runner util.CmdRunner,
+) (Runtime, error) {
+	err := util.ValidateDockerTools(runner)
```
Will this be the code that eventually switches between Docker+Firecracker?

Piszmog replied:
Yes! all the runtimes will be here - Firecracker, docker, K8s, whatever

eseliger commented on enterprise/cmd/executor/singlebinary/doc.go:
```diff
@@ -1,3 +1,3 @@
-// singlebinary is the entrypoint for the sourcegraph single binary process. It must
+// Package singlebinary is the entrypoint for the sourcegraph single binary process. It must
 // not be used or imported from anywhere else.
```
oh interesting, I did not know this convention. TIL!

eseliger commented on enterprise/cmd/executor/internal/run/util.go:
```diff
@@ -16,14 +14,14 @@ import (
 )
 
-func newQueueTelemetryOptions(ctx context.Context, useFirecracker bool, logger log.Logger) queue.TelemetryOptions {
+func newQueueTelemetryOptions(runner util.CmdRunner, ctx context.Context, useFirecracker bool, logger log.Logger) queue.TelemetryOptions {
```
I think context should always come first in function signatures

eseliger replied:
```suggestion
func newQueueTelemetryOptions(ctx context.Context, runner util.CmdRunner, useFirecracker bool, logger log.Logger) queue.TelemetryOptions {
```

eseliger commented on enterprise/cmd/executor/internal/run/install.go:
```diff
@@ -291,19 +292,19 @@ func installCNIPlugins(cliCtx *cli.Context) error {
 }
 
-func installSrc(cliCtx *cli.Context, logger log.Logger, config *config.Config) error {
+func installSrc(runner util.CmdRunner, cliCtx *cli.Context, logger log.Logger, config *config.Config) error {
```
same here, I think context should always be first (yes this is cliCtx but it also contains the ctx)

```suggestion
func installSrc(cliCtx *cli.Context, runner util.CmdRunner, logger log.Logger, config *config.Config) error {
```

eseliger commented on enterprise/cmd/executor/internal/util/validate.go:
```diff
@@ -0,0 +1,223 @@
+	if err != nil {
+		return errors.Newf("failed to semver parse git version: %s", gitVersion)
+	} else if !config.MinGitVersionConstraint.Check(have) {
+		return errors.Newf("git version is too old, install at least git 2.26, current version: %s", gitVersion)
```
lol I just realize that we don't print MinGitVersionConstraint in here but a hardcoded 2.26. Is it easy to stringify a semver?

Piszmog replied:
Yea, I'll update for `MinGitVersionConstraint`

eseliger commented on enterprise/cmd/executor/internal/util/validate.go:
```diff
@@ -0,0 +1,223 @@
+}
+
+// ValidateFirecrackerTools validates that the tools required to run Firecracker are installed.
+func ValidateFirecrackerTools(runner CmdRunner) error {
```
This might be something we want to expose on the runtime interface later (not this PR), too: `ValidateTools() error` so this is not mixing different runtimes in here.

eseliger commented on enterprise/cmd/executor/internal/worker/workspace.go:
```diff
@@ -18,21 +18,16 @@ func (h *handler) prepareWorkspace(
 	commandLogger command.Logger,
 ) (workspace.Workspace, error) {
-	if h.options.FirecrackerOptions.Enabled {
+	if h.options.CommandOptions.FirecrackerOptions.Enabled {
```
Potentially also something we could move on the runtime interface: `CreateWorkspace`

eseliger commented on enterprise/cmd/executor/internal/worker/runtime/runtime.go:
```diff
@@ -0,0 +1,57 @@
+	cloneOpts workspace.CloneOptions,
+	runner util.CmdRunner,
+) (Runtime, error) {
+	err := util.ValidateDockerTools(runner)
+	if err != nil {
+		var errMissingTools *util.ErrMissingTools
+		if errors.As(err, &errMissingTools) {
+			logger.Warn("runtime 'docker' is not supported: missing required tools", log.Strings("dockerTools", errMissingTools.Tools))
+		} else {
+			logger.Warn("failed to determine if docker tools are configured", log.Error(err))
+		}
```
I wonder if this should be exposed as a method on the runtime 

```
type Runtime interface { ValidateInstallation(Runner) error }

func New(type RuntimeType) {
switch type {
case RuntimeTypeDocker:
r := docker.NewRuntime()
r.ValidateTools()
return r
case RuntimeTypeFirecracker:
r := firecracker.NewRuntime()
r.ValidateTools()
return r
}
}
```

Piszmog replied:
Yea, I was thinking this too. As I was working this, I did write up an Issue (#47819) to expose some runtime stuff in the CLI. This would prob be a good candidate

eseliger reviewed (approved):
Great stuff, really really impressed with the amount of testing you've put in here and how things are starting to pan out. This already feels a good bunch cleaner :) 

> Since this is a first pass with implementing a single runtime, things feel squishy. Between trying to stay backwards compatible and not knowing what I don't know, it's hard to get a feel for how the APIs/interfaces should be. I feel this will come together more once the other runtimes are worked on.

Fully onboard with that, happy to merge this as-is and let it pan out more as we implement more runtimes and learn more :)

Piszmog commented:
@eseliger per the squishiness, last couple days I felt meh about this implementation. So I have recently started to fix some the squishy feelings I have. I'll comment when I have pushed that work. 

Essentially we have a relationship between `runtime` and `command` (runners and the command), and I want to make clear the separation of concerns and make it easier to integrate the runners into the runtimes. So I'm splitting things apart and see how it pans out. So far, it feel better to me and will be easier to integrate into new runtimes like K8s.

eseliger commented:
alright! curious what you came up with :)

Piszmog commented:
So I pushed the changes. Don't get freaked out by the diff (lol). Gist of changes,

- Moved `command` into the `worker` package
- Split runners and commands into separate packages
    - Runners use commands to execute their stuff
- Full test coverage of changes 

Lots of stuff is still borked, but the general changes have been pushed

Piszmog commented:
Ok, this PR is pretty much done. Now you may be thinking (like me). Holy @#$! 11k new lines!!!!

80% of these new lines are for unit tests. I added over 120 new unit tests

Piszmog commented:
Pushed these changes to [executor-patch-notest/runtime-interface](https://buildkite.com/sourcegraph/sourcegraph/builds?branch=executor-patch-notest%2Fruntime-interface) to get an image built without merging to `main`

sanderginn commented on enterprise/cmd/executor/internal/worker/command/command.go:
```diff
@@ -145,39 +167,41 @@ func prepCommand(ctx context.Context, command command) (cmd *exec.Cmd, stdout, s
 // we shell out to, such a docker.
```
What does this comment mean 😂

Piszmog replied:
we 🐚

Piszmog replied:
I'll fix

sanderginn commented on enterprise/cmd/executor/internal/worker/handler.go:
```diff
@@ -80,65 +86,138 @@ func (h *handler) Handle(ctx context.Context, logger log.Logger, job types.Job)
+	return nil
+}
+
+func createHoneyEvent(_ context.Context, job types.Job, err error, duration time.Duration) honey.Event {
+	fields := map[string]any{
+		"duration_ms":    duration.Milliseconds(),
+		"recordID":       job.RecordID(),
+		"repositoryName": job.RepositoryName,
+		"commit":         job.Commit,
+		"numDockerSteps": len(job.DockerSteps),
+		"numCliSteps":    len(job.CliSteps),
+	}
+
+	if err != nil {
+		fields["error"] = err.Error()
+	}
+
+	return honey.NewEventWithFields("executor", fields)
+}
```
This seems oddly specific to our internal systems 🤔

Piszmog replied:
I know nothing about this 🍯 . Just moved it from elsewhere

sanderginn replied:
@eseliger any idea? Not everyone uses Honeycomb?

sanderginn commented on enterprise/cmd/executor/internal/worker/runner/docker.go:
```diff
@@ -0,0 +1,93 @@
+	}
+}
+
+func (r *dockerRunner) TempDir() string {
+	return r.tmpDir
+}
```
What's the advantage of making this a func over exporting `tmpDir`?

Piszmog replied:
Unit testing. I generally like blackbox testing. This means my test files are `package name_test`.

So in the Unit Tests we call `NewDockerRunner`. But this returns the interface `Runner`. So we wont have access to this property in the tests. So the function is just for unit testing (see the comment on `Runner.TempDir()`)

sanderginn replied:
I saw the interface after posting this comment, but I figured I'd leave the comment because I was interested to see what you'd have to share. Makes sense!

sanderginn commented on enterprise/cmd/executor/internal/worker/runner/docker.go:
```diff
@@ -0,0 +1,93 @@
+
+	// If docker auth config is present, write it.
+	if len(r.dockerAuthConfig.Auths) > 0 {
+		d, err := json.Marshal(r.dockerAuthConfig)
+		if err != nil {
+			return err
+		}
```
The marshalled version is already present in `config.DockerAuthConfigStr`. Could that be used here instead?

Piszmog replied:
Actually this is slightly convoluted 🙂 

A user can configured a DockerAuth config for the whole of an executor instance. But a `Job` may come with it's own auth specified. In that case we will use the job auth over the executor instance.

Now as I doubled checked this, I found that my changes removed this logic. So adding it back 😄  (+ unit tests)

sanderginn commented on enterprise/cmd/executor/internal/worker/runner/firecracker.go:
```diff
@@ -0,0 +1,345 @@
+	// If docker auth config is present, write it.
+	var dockerConfigPath string
+	if len(r.dockerAuthConfig.Auths) > 0 {
+		d, err := json.Marshal(r.dockerAuthConfig)
+		if err != nil {
+			return "", err
+		}
```
Same comment as with the Docker runner

sanderginn commented on enterprise/cmd/executor/internal/worker/runner/firecracker.go:
```diff
@@ -0,0 +1,345 @@
+	if err != nil {
+		return "", err
+	}
+	cniConfigFile := path.Join(cniConfigDir, "10-sourcegraph-executors.conflist")
```
What's the significance of the `10-`?

Piszmog replied:
Good question 🤔 

@eseliger what is the significance of this? (I'll add a comment with the _why_)

Piszmog commented on enterprise/cmd/executor/internal/worker/command/command.go:
```diff
@@ -116,16 +121,25 @@ func validateCommand(command []string) error {
 	}
 
-	cmd.Env = env
+	cmd.Env = append(cmd.Env, env...)
```
Yea not sure what my brain was think. I changed this back to `cmd.Env = env`

Piszmog replied:
Now I remember why. This was allowing me to inject the ENV vars when running the unit test with the tricky cmd.Exec

sanderginn reviewed (approved):
I really like the patterns you implemented here. Awesome job 👍 

Most of my comments are actually questions I have to understand executors better. A few nits otherwise.

sanderginn commented on enterprise/cmd/executor/internal/run/util.go:
```diff
@@ -141,21 +102,23 @@ func dockerOptions(c *config.Config) command.DockerOptions {
 		SandboxImage:             c.FirecrackerSandboxImage,
 		VMStartupScriptPath:      c.VMStartupScriptPath,
 		DockerRegistryMirrorURLs: dockerMirrors,
+		DockerOptions:            dockerOptions(c),
```
I guess to pull Docker images lol

Piszmog commented on enterprise/cmd/executor/internal/run/install.go:
```diff
@@ -15,32 +15,33 @@ import (
 )
 
-func RunInstallIgnite(cliCtx *cli.Context, logger log.Logger, config *config.Config) error {
+func InstallIgnite(cliCtx *cli.Context, runner util.CmdRunner, logger log.Logger, config *config.Config) error {
```
`runner` has to be passed as it is being forced by `makeActionHandler` in `main.go`

sanderginn replied:
Ah, ack

Piszmog commented on enterprise/cmd/executor/internal/run/install.go:
```diff
@@ -15,32 +15,33 @@ import (
 }
 
-func RunInstallIPTablesRules(cliCtx *cli.Context, logger log.Logger, config *config.Config) error {
+func InstallIPTablesRules(cliCtx *cli.Context, runner util.CmdRunner, logger log.Logger, config *config.Config) error {
```
Same - it is forced by `main.go`

Piszmog commented on enterprise/cmd/executor/internal/run/run.go:
```diff
@@ -14,17 +14,18 @@ import (
 
-func RunRun(cliCtx *cli.Context, logger log.Logger, cfg *config.Config) error {
-	return StandaloneRunRun(cliCtx.Context, logger, cfg, cliCtx.Bool("verify"))
+func Run(cliCtx *cli.Context, runner util.CmdRunner, logger log.Logger, cfg *config.Config) error {
+	return StandaloneRun(cliCtx.Context, runner, logger, cfg, cliCtx.Bool("verify"))
```
lol sames

Piszmog commented on enterprise/cmd/executor/internal/run/run.go:
```diff
@@ -128,12 +130,12 @@ func StandaloneRunRun(ctx context.Context, logger log.Logger, cfg *config.Config
 }
 
-func mustRegisterVMCountMetric(logger log.Logger, observationCtx *observation.Context, prefix string) {
+func mustRegisterVMCountMetric(runner util.CmdRunner, logger log.Logger, observationCtx *observation.Context, prefix string) {
```
whoa, don't shoot me. I didn't start it

Piszmog commented on enterprise/cmd/executor/internal/run/testvm.go:
```diff
@@ -61,10 +65,13 @@ func createVM(ctx context.Context, config *config.Config, repositoryName, revisi
-	hostRunner := command.NewRunner("", commandLogger, command.Options{}, operations)
+	cmd := &command.RealCommand{
+		CmdRunner: cmdRunner,
+		Logger:    log.Scoped("executor-test-vm", ""),
```
So `command.Logger` is different from `log.Logger` (command logger doesn't implement logger). `Logger` is for "internal" logging vs command logger is for _potentially_ writing the logs back to sourcegraph. So we couldn't pass the command logger to this field.

sanderginn replied:
Gotcha!

Piszmog commented on enterprise/cmd/executor/internal/util/validate.go:
```diff
@@ -0,0 +1,223 @@
+}
+
+// ValidateCNIInstalled validate that the CNI plugins for firecracker are properly installed.
+func ValidateCNIInstalled() error {
```
This comes in the next PR (firecracker PR - #48752) 😄

sanderginn replied:
Lol oops. You're putting out too much work to keep track of man!

Piszmog commented on enterprise/cmd/executor/internal/util/validate.go:
```diff
@@ -0,0 +1,223 @@
+  $ curl -sSL https://github.com/containernetworking/plugins/releases/download/v0.9.1/cni-plugins-linux-amd64-v0.9.1.tgz | tar -xz -C /opt/cni/bin`
+		if missingIsolationPlugin {
+			hint += `
+  $ curl -sSL https://github.com/AkihiroSuda/cni-isolation/releases/download/v0.0.4/cni-isolation-amd64.tgz | tar -xz -C /opt/cni/bin`
```
Looks like that repo is archived. So I wouldn't expect anymore releases coming from here.

But the repo says it has been merged into CNI firewall plugin. I do not know what consequences that has for us. I don't think we should maintain our own fork.

Piszmog commented on enterprise/cmd/executor/internal/util/validate.go:
```diff
@@ -0,0 +1,223 @@
+// ValidateFirecrackerTools validates that the tools required to run Firecracker are installed.
+func ValidateFirecrackerTools(runner CmdRunner) error {
+	var missingTools []string
+	for _, tool := range config.RequiredCLIToolsFirecracker {
```
I checked all the consts and most are used in the `util` package, some in the `worker` package and a couple in the `config` package. 

I am mixed on moving the values to closer where they are used or keeping as is (if they span multiple packages, then some where else to avoid cyclical dependency).

My other thought a lot of this is like infra type things. So it might be good to keep them al together (a couple are questionable from my perspective). So in a sense they are "configurations" (making sure proper versions are installed or certain tools are available) but more "infrastructure".

My gut, is to move the consts closer to usage. They consts that are related to one another would go together and used in same-ish places.

Piszmog commented on enterprise/cmd/executor/internal/util/tools.go:
```diff
@@ -0,0 +1,34 @@
+)
+
+// GetGitVersion returns the version of git installed on the host.
+func GetGitVersion(ctx context.Context, runner CmdRunner) (string, error) {
+	out, err := execOutput(ctx, runner, "git", "version")
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimPrefix(out, "git version "), nil
+}
+
+// GetSrcVersion returns the version of src installed on the host.
+func GetSrcVersion(ctx context.Context, runner CmdRunner) (string, error) {
+	out, err := execOutput(ctx, runner, "src", "version", "-client-only")
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimPrefix(out, "Current version: "), nil
+}
+
+// GetDockerVersion returns the version of docker installed on the host.
+func GetDockerVersion(ctx context.Context, runner CmdRunner) (string, error) {
+	return execOutput(ctx, runner, "docker", "version", "-f", "{{.Server.Version}}")
+}
+
+// GetIgniteVersion returns the version of ignite installed on the host.
+func GetIgniteVersion(ctx context.Context, runner CmdRunner) (string, error) {
+	return execOutput(ctx, runner, "ignite", "version", "-o", "short")
+}
```
I might be misunderstanding, but these versions are being used for validating that the right "thing" is installed. So for example, the value of `GetSrcVersion` is used for telemetry and min version checking. It's not directly exposed to the user. And when it is, we put it more into a sentence (e.g. "You have 4.1 installed of src-cli but you really want 5.1")

Piszmog commented:
If there are no more comments, I will merge this tomorrow (one more day of running in dogfood)

sanderginn commented:
All good from me 👍
